src/algebra/lie_algebras.lean

```diff
@@ -0,0 +1,192 @@
+
+namespace lie_ring
+
+variables (L : Type v) [add_comm_group L] [lie_ring L]
+
+attribute [simp] add_left
+attribute [simp] add_right
+attribute [simp] alternate
+attribute [simp] jacobi
+
+@[simp] protected lemma skew (x y : L) :
+  -⁅y, x⁆ = ⁅x, y⁆ :=
+begin
+  symmetry,
+  rw [←sub_eq_zero_iff_eq, sub_neg_eq_add],
+  have H : ⁅x + y, x + y⁆ = 0 := by apply lie_ring.alternate,
+  rw [lie_ring.add_left,
+      lie_ring.add_right, lie_ring.add_right,
+      lie_ring.alternate, lie_ring.alternate] at H,
+  simp at H,
+  exact H,
+end
+
+@[simp] protected lemma zero_left (x : L) :
+  ⁅x, 0⁆ = 0 :=
+begin
+  have H : ⁅x, 0⁆ + ⁅x, 0⁆ = ⁅x, 0⁆ + 0 := by { rw ←lie_ring.add_right, simp, },
+  exact add_left_cancel H,
+end
+
+@[simp] protected lemma zero_right (x : L) :
+  ⁅0, x⁆ = 0 := by { rw [←lie_ring.skew, lie_ring.zero_left], simp, }
+
+@[simp] protected lemma neg_left (x y : L) :
+  ⁅-x, y⁆ = -⁅x, y⁆ := by { rw [←sub_eq_zero_iff_eq, sub_neg_eq_add, ←lie_ring.add_left], simp, }
+
+@[simp] protected lemma neg_right (x y : L) :
+  ⁅x, -y⁆ = -⁅x, y⁆ := by { rw [←lie_ring.skew, ←lie_ring.skew], simp, }
+
+@[simp] protected lemma gsmul_left (x y : L) (n : ℤ) :
+  ⁅n • x, y⁆ = n • ⁅x, y⁆ :=
+begin
+  have H : ∀ (m : ℕ), ⁅m • x, y⁆ = m • ⁅x, y⁆ := by {
+    intros, induction m with m IHm,
+    { simp, },
+    { rw [←nat.one_add, add_smul, lie_ring.add_left, IHm, add_smul], simp, }, },
+  cases n,
+  { apply H, },
+  { unfold has_scalar.smul, unfold has_scalar.smul at H,
+    rw [gsmul_neg_succ, gsmul_neg_succ, lie_ring.neg_left, H], },
+end
+
+@[simp] protected lemma gsmul_right (x y : L) (n : ℤ) :
+  ⁅x, n • y⁆ = n • ⁅x, y⁆ :=
+begin
+  rw [←lie_ring.skew, ←lie_ring.skew _ x, lie_ring.gsmul_left],
+  unfold has_scalar.smul, rw gsmul_neg,
+end
+
+instance of_associative_ring (A : Type v) [ring A] : lie_ring A :=
+{ add_left  := ring_commutator.add_left A,
+  add_right := ring_commutator.add_right A,
+  alternate := ring_commutator.alternate A,
+  jacobi    := ring_commutator.jacobi A }
+
+end lie_ring
+
+/--
+A Lie algebra is a module with compatible product, known as the bracket, satisfying the Jacobi
+identity. Forgetting the scalar multiplication, every Lie algebra is a Lie ring.
+-/
+class lie_algebra (R : Type u) (L : Type v)
+  [comm_ring R] [add_comm_group L] extends module R L, lie_ring L :=
+(smul_right : ∀ (t : R) (x y : L), ⁅x, t • y⁆ = t • ⁅x, y⁆)
+
+namespace lie_algebra
+
+variables (R : Type u) (L : Type v) [comm_ring R] [add_comm_group L] [lie_algebra R L]
+
+attribute [simp] smul_right
+
+@[simp] protected lemma smul_left (t : R) (x y : L) :
+  ⁅t • x, y⁆ = t • ⁅x, y⁆ :=
+by { rw [←lie_ring.skew, ←lie_ring.skew _ x y], simp [-lie_ring.skew] }
+
+def adjoint_action (x : L) : L →ₗ[R] L :=
+{ to_fun := has_bracket.bracket x,
+  add    := by { intros, apply lie_ring.add_right },
+  smul   := by { intros, apply lie_algebra.smul_right } }
+
+def bracket_bilinear : L →ₗ[R] L →ₗ[R] L :=
```

Comment:
This name is somewhat long. Sometimes people have prefixed the unbundled name with an `l`, for `linear`. So this might be `lAd`. Admittedly, a very ugly name. Another option: `bil_lie`?